core/drivers/clk/clk_of.c

```diff
@@ -0,0 +1,368 @@
+					void *data)
+{
+	struct clk_of_provider *prv = NULL;
+
+	prv = malloc(sizeof(*prv));
+	if (!prv)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	prv->get_of_clk = get_of_clk;
+	prv->data = data;
+	prv->nodeoffset = nodeoffset;
+	prv->clock_cells = fdt_clock_cells(fdt, nodeoffset);
+	prv->phandle = fdt_get_phandle(fdt, nodeoffset);
+
+	SLIST_INSERT_HEAD(&clk_of_provider_list, prv, link);
+
+	return TEE_SUCCESS;
+}
+
+static struct clk_of_provider *clk_get_provider(int nodeoffset)
+{
+	struct clk_of_provider *prv = NULL;
+
+	SLIST_FOREACH(prv, &clk_of_provider_list, link) {
+		if (prv->nodeoffset == nodeoffset)
+			return prv;
+	}
+
+	return NULL;
+}
+
+static struct clk_of_provider *clk_get_provider_by_phandle(uint32_t phandle)
+{
+	struct clk_of_provider *prv = NULL;
+
+	SLIST_FOREACH(prv, &clk_of_provider_list, link) {
+		if (prv->phandle == phandle)
+			return prv;
+	}
+
+	return NULL;
+}
+
+int clk_of_get_idx_by_name(void *fdt, int nodeoffset, const char *name)
+{
+	int idx = -1;
+
+	idx = fdt_stringlist_search(fdt, nodeoffset, "clock-names", name);
+	if (idx < 0)
+		return -1;
+
+	return idx;
+}
+
+static struct clk *clk_of_get_from_provider(struct clk_of_provider *prv,
+					    unsigned int clock_cells,
+					    const uint32_t *prop)
+{
+	unsigned int arg = 0;
+	struct clk_of_phandle_args pargs = {0};
+
+	pargs.args_count = clock_cells;
+	for (arg = 0; arg < clock_cells; arg++)
+		pargs.args[arg] = fdt32_to_cpu(prop[arg + 1]);
+
+	return prv->get_of_clk(&pargs, prv->data);
+}
+
+struct clk *clk_of_get_by_name(void *fdt, int nodeoffset,
+			       const char *name)
+{
+	int clk_id = 0;
+
+	clk_id = clk_of_get_idx_by_name(fdt, nodeoffset, name);
+	if (clk_id < 0)
+		return NULL;
+
+	return clk_of_get_by_idx(fdt, nodeoffset, clk_id);
+}
+
+static struct clk *clk_of_get_by_idx_prop(const char *prop_name, void *fdt,
+					  int nodeoffset, int clk_idx)
+{
+	int len = 0;
+	int idx = 0;
+	int idx32 = 0;
+	int clock_cells = 0;
+	uint32_t phandle = 0;
+	const uint32_t *prop = NULL;
+	struct clk_of_provider *prv = NULL;
+
+	prop = fdt_getprop(fdt, nodeoffset, prop_name, &len);
+	if (!prop)
+		return NULL;
+
+	while (idx < len) {
+		idx32 = idx / sizeof(uint32_t);
+		phandle = fdt32_to_cpu(prop[idx32]);
+
+		prv = clk_get_provider_by_phandle(phandle);
+		if (!prv)
+			return NULL;
+
+		clock_cells = prv->clock_cells;
+		if (clk_idx == 0)
+			return clk_of_get_from_provider(prv,
+							clock_cells,
+							&prop[idx32]);
+
+		clk_idx--;
+		idx += sizeof(phandle) + clock_cells * sizeof(uint32_t);
+	}
+
+	return NULL;
+}
+
+struct clk *clk_of_get_by_idx(void *fdt, int nodeoffset, int clk_idx)
+{
+	return clk_of_get_by_idx_prop("clocks", fdt, nodeoffset, clk_idx);
+}
+
+static const struct clk_driver *clk_get_compatible_driver(const char *compat)
+{
+	const struct dt_driver *drv = NULL;
+	const struct dt_device_match *dm = NULL;
+	const struct clk_driver *clk_drv = NULL;
+
+	for_each_dt_driver(drv) {
+		if (drv->type != DT_DRIVER_CLK)
```

Comment:
here also i think `compatible` is sufficient to find the right driver.

Reply:
Actually, the for_each_dt_driver iterate on all drivers registered in the dt_driver section (which almso contain UART drivers). So we might end up casting the uart driver field into a clk driver field and call probe function.